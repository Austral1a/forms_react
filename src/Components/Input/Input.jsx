import React, {useEffect, useMemo, useCallback} from 'react'
import classNames from 'classnames'
import './Input.scss'

export const Input = React.forwardRef((
    {
        type,
        placeholder,
        onChange,
        onInput,
        onBlur,
        value,
        hasError,
        errorMessage,
        disabled,
        icon,
        onFocus,
        className
    },
    ref,
) => {
    const customClasses = classNames('input-container', className)
    return(
        <div className={customClasses}>
            {icon && <span className='input-container__icon'><img src={icon}/></span>}
            <input
                ref={ref}
                type={type}
                placeholder={placeholder}
                className='input-container__input'
                onChange={onChange}
                onInput={onInput}
                onBlur={onBlur}
                value={value}
                style={{
                    borderBottom: hasError && !!value ? '2px solid red' : 'inherit'
                }}
                disabled={disabled}
                onFocus={onFocus}
            />
            {hasError && !!value && <p className='input-container__error'>{errorMessage}</p>}
        </div>
    )
})
