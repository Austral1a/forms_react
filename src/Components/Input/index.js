export * from './Input.scss'
export * from './Input'
