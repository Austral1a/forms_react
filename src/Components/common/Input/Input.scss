.input-container {
  display: grid;
  grid-template-columns: 1fr;
  grid-template-rows: repeat(2, 30px);
  position: relative;

  .input-container__input {
    padding: 3px 0 3px 25px;
    border-radius: 4px;
  }

  .input-container__icon {
    img {
      width: 20px !important;
      height: 20px !important;
      opacity: 0.8;
    }
    position: absolute;
    top: 5px;
    left: 5px;
  }
}
