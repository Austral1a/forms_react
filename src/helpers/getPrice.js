const getPrice = (subsc, country, plan) => {
    return subsc[country][plan]
}

export default getPrice
