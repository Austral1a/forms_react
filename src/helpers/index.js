export * from './validationMessages'
export * from './patterns'
export * from './subscriptions'
export * from './masks'
export * from './getPrice'
export * from './isInputValid'
export * from './translations'
