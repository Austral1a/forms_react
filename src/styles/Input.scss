.input-container {
  display: grid;
  grid-template-columns: 1fr;
  grid-template-rows: repeat(2, 30px);

  .input-container__input {
    padding: 4px 7px;
    border-radius: 4px;
  }
}
