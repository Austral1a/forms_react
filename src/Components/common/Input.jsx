import React, {useEffect, useMemo, useCallback} from 'react'

import '../../styles/Input.scss'

const Input = React.forwardRef((
    {
        type,
        placeholder,
        onChange,
        onInput,
        onBlur,
        value,
        hasError,
        errorMessage,
        disabled,
        icon
    },
    ref,
) => {


    const renderError = useCallback(() => {
        console.log(hasError, value)
        if(!hasError) {
            return null
        } else if(value === '') {
            return ''
        } else if(hasError) {
            return <p className='input-container__error'>{errorMessage}</p>
        }
    }, [hasError, value])

    return(
        <div className='input-container'>
            {icon ? <span className='input-container__icon'><img src={icon}/></span> : null}
            <input
                ref={ref}
                type={type}
                placeholder={placeholder}
                className='input-container__input'
                onChange={onChange}
                onInput={onInput}
                onBlur={onBlur}
                value={value}
                style={{
                    borderBottom: hasError && value !== '' ? '2px solid red' : 'inherit'
                }}
                disabled={disabled}
            />
            {renderError()}
        </div>
    )
})

export default Input
