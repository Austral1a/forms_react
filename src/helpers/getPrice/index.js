import {subscriptions as sub} from '../subscriptions/subscriptions'
import getPrice from './getPrice'
export {
    sub,
    getPrice
}
