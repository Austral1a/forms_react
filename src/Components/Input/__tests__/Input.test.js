import {Input} from '../Input'
import {shallow} from 'enzyme'
import toJson from 'enzyme-to-json'
import {isInputValid, patterns} from '../../../helpers'
describe('Input component', function () {

    let wrapper
    const errorMsg = 'Error has occurred'

    beforeEach(() => {
        wrapper = shallow(<Input />)
    })

    it('Input must match snapshot', () => {

        expect(toJson(wrapper)).toMatchSnapshot()
    })

    it("renders '.input-container'", () => {

        expect(wrapper.find('.input-container')).toHaveLength(1)
    })

    it("renders '.input-container__icon' when icon prop is exists", () => {
        wrapper = shallow(<Input icon='path/to/icon' />)
        expect(wrapper.find('.input-container__icon')).toHaveLength(1)
    })

    it("not renders '.input-container__icon' when icon prop is not exists", () => {
        expect(wrapper.find('.input-container__icon')).toHaveLength(0)
    })

    it("must correctly handle 'type' prop", () => {
        const typeText = shallow(<Input type='text' />)
        expect(typeText.find('input').prop('type')).toBe('text')

        const typePassword = shallow(<Input type='password' />)
        expect(typePassword.find('input').prop('type')).toBe('password')
    })

    it("must correctly handle 'placeholder' prop", () => {
        const firstName = shallow(<Input placeholder='First Name' />)
        const lastName = shallow(<Input placeholder='Last Name' />)
        const email = shallow(<Input placeholder='Email' />)

        expect(firstName.find('input').prop('placeholder')).toBe('First Name')
        expect(lastName.find('input').prop('placeholder')).toBe('Last Name')
        expect(email.find('input').prop('placeholder')).toBe('Email')
    })

    it("must correctly handle 'value' prop", () => {
        const kyle = shallow(<Input value='Kyle' />)
        const kyl3 = shallow(<Input value='kyl3' />)

        expect(kyle.find('input').prop('value')).toBe('Kyle')
        expect(kyl3.find('input').prop('value')).toBe('kyl3')
    })

    describe("must correctly handle 'hasError' prop", () => {
        const wrapperWithError = shallow(<Input
            hasError={true}
            errorMessage={errorMsg}
            value='some not g00d value'
        />)
        const wrapperWithNoError = shallow(<Input
            hasError={false}
            errorMessage={errorMsg}
            value='some good value'
        />)

        it("must show error message if 'hasError' prop is true", () => {

            expect(wrapperWithError.find('.input-container__error')).toHaveLength(1)
            expect(wrapperWithError.find('.input-container__error').text()).toBe(errorMsg)
        })

        it("must not show error message if 'hasError' prop is false", () => {

            expect(wrapperWithNoError.find('.input-container__error')).toHaveLength(0)
        })
    })

    it("must correctly handle 'onChange' prop", () => {
        const newValue = 'new value'
        const state = {
            value: ''
        }
        const handleChange = (e) => {
            state['value'] += e.target.value
        }
        const wrapper = shallow(<Input onChange={handleChange} />)
        wrapper.find('.input-container__input').simulate('change', {
            target: {
                value: newValue
            }
        })
        expect(state.value).toBe(newValue)
    })

    describe("must correctly handle 'onBlur' prop", () => {
        let state
        let handleBlur

        beforeEach(() => {
            state = {
                hasError: false
            }
            handleBlur = (pattern, text) => {
                state.hasError = !isInputValid(patterns[pattern], text);
            }
        })
        let wrapper
        const middleEach = (inputVal, pattern) => {
            wrapper = shallow(<Input
                hasError={state.hasError}
                value='some value'
                errorMessage={errorMsg}
                onBlur={() => handleBlur(pattern, inputVal)} />)
            wrapper.find('.input-container__input').simulate('blur')
            wrapper.setProps({hasError: state.hasError})
        }


        describe('validator pattern is name', () => {

            it("input val 'Kyle2' with name pattern validation must render error", () => {
                const inputVal = 'Kyle2'

                middleEach(inputVal, 'name')
                expect(wrapper.find('.input-container__error')).toHaveLength(1)
                expect(wrapper.find('.input-container__error').text()).toBe(errorMsg)
            })

            it("input val 'kyle' with name pattern validation must not render an error", () => {
                const inputVal = 'kyle'

                middleEach(inputVal, 'name')
                expect(wrapper.find('.input-container__error')).toHaveLength(0)
            })
        })


        describe('validator pattern is email',() => {

            it("input val 'email@em.com' must not render error", () => {
                const inputVal = 'email@em.com'

                middleEach(inputVal, 'email')
                expect(wrapper.find('.input-container__error')).toHaveLength(0)

            })

            it("input val 'email@.com must render an error'", () => {
                const inputVal = 'email@.com'

                middleEach(inputVal, 'email')
                expect(wrapper.find('.input-container__error')).toHaveLength(1)
                expect(wrapper.find('.input-container__error').text()).toBe(errorMsg)
            })

            it("input val 'emal@com' must render an error", () => {
                const inputVal = 'emal@com'

                middleEach(inputVal, 'email')
                expect(wrapper.find('.input-container__error')).toHaveLength(1)
                expect(wrapper.find('.input-container__error').text()).toBe(errorMsg)
            })
        })

        describe('validator pattern is phone', () => {

            it("input val '(123) 456 - 7890' must not render an error", () => {
                const inputVal = '(123) 456 - 7890'

                middleEach(inputVal, 'phone')
                expect(wrapper.find('.input-container__error')).toHaveLength(0)
            })

            it("input val '123 456 - 7890' must render an error", () => {
                const inputVal = '123 456 - 7890'

                middleEach(inputVal, 'phone')
                expect(wrapper.find('.input-container__error')).toHaveLength(1)
                expect(wrapper.find('.input-container__error').text()).toBe(errorMsg)
            })


        })
    })
})
