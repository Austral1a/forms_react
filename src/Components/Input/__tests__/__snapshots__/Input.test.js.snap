// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component Input must match snapshot 1`] = `
<div
  className="input-container"
>
  <input
    className="input-container__input"
  />
</div>
`;
